### PR TITLE
Make getUserObject* return the correct thread copy for MaterialUserObjects 

### PR DESCRIPTION
**Pull request by dschwen**
This obviates the need for `getMaterialUserObject` and makes this type of user object available in other places than just Materials. We'll stick to the name anyhow, as the UO type acts as a replacement for MaterialProperties in its main usecase.

Refs #6233 (<s>this PR contains the commit from PR #6262</s> merged into devel and rebased)

### CODE REVIEW

**Comment by friedmud** [comment]
"its main usecase"... "that I can think of (and I didn't try very hard)". ;-)

MaterialUserObject is just not the right word for this at all.  People are going to use it for MUCH more than that... you only think it should be called that because that's the purpose you originally thought of it for.  It would be like naming DiracKernels "PointSources" or naming AuxKernels "VelocityCalculators" (that's what it was originally used for), etc.

We need to come up with a name that is descriptive and generic.  "MaterialUserObject" says nothing about what the object does (I would actually assume, based on our UserObject naming scheme, that it's an intermediate base class that you inherit from to create UserObjects that can _use_ Material properties!!!!!).

Further... this thing inherits from ElementUserObject... But the name doesn't reflect the fact that it will only work on Elements at all (in fact, how are you guys going to deal with that because we need to evaluate Material Properties on _sides_ as well.. So how does that work with this?).

So... Let's look at these things they are:

1.  Elemental
2.  Local (I.e. They don't do global reductions like most of the UserObjects), they just work "locally" on one element.

I would call them `LocalElementUserObjects`.

Regardless of if you like that name or not... `MaterialUserObject` is not right.

**Comment by dschwen** [comment]
Ok

**Comment by dschwen** [comment]
Yeah, I like this name much better. This will allow us to add a `LocalSideUserObject` at some point (I should probably rename `isLocalElementUserObject` to `isLocalUserObject` and add further tests in the when necessary).

**Comment by friedmud** [comment]
The only issue I have with "local" is that it's also used to mean "on processor".... as in "mesh.local_elements".

Do you think it's going to cause much confusion?  It could lead people to believe that these are UserObjects that only act on the "local" processor.

...

**Comment by dschwen** [comment]
Hmm, good point. But the user object is effectively only being called on the local process and thread. The threadJoin and finalize methods are never called :-)

**Comment by friedmud** [comment]
Yes, _we_ know that to be true.  I'm just talking about possible confusion for users.

We could do: `IndividualElementUserObject` or `SingleElementUserObject` ?  That last one has a decent ring to it...

**Comment by dschwen** [comment]
Sure `SingleElementUserObject` sounds good. But then my tester function in UOI would be named `isSingleUserObject`.

```
if (isSingleUserObject(uo))
  uo.putARingOnIt();
```

**Comment by friedmud** [comment]
That is kind of odd...

`DiscreteElementUserObject`?

**Comment by dschwen** [comment]
Yeah, ok. Discrete is fine. It is in fact so discrete that is does not tattle its data to other threads and processors ;-)

Before I co ahead and change the name again in two PRs and one branch in a separate app, is there anything else we need to fix?

File: framework/include/userobject/UserObjectInterface.h [hunk 1]
```diff
@@ -84,14 +87,16 @@ template<class T>
 const T &
 UserObjectInterface::getUserObject(const std::string & name)
 {
-  return _uoi_feproblem.getUserObject<T>(_uoi_params.get<UserObjectName>(name));
+  unsigned int tid = isLocalUserObject(getUserObjectBase(name)) ? _uoi_tid : 0;
```
Comment (friedmud) [inline framework/include/userobject/UserObjectInterface.h]: Is it not feasible to do this inside the Warehouse itself?

File: framework/src/utils/ParallelUniqueId.C [hunk 2]
```diff
@@ -18,7 +18,7 @@ bool ParallelUniqueId::_initialized = false;
 
 #ifdef LIBMESH_HAVE_TBB_API
 tbb::concurrent_bounded_queue<unsigned int> ParallelUniqueId::_ids;
-#elif LIBMESH_HAVE_PTHREAD
+#elif !defined(LIBMESH_HAVE_OPENMP) && defined(LIBMESH_HAVE_PTHREAD)
```
Comment (friedmud) [inline framework/src/utils/ParallelUniqueId.C]: What's this doing here?  Seems unrelated...

Comment (dschwen) [inline framework/src/utils/ParallelUniqueId.C]: This will be gone when reading. It is a commit from a PR that was merged. See the original PR description.

File: framework/include/userobject/UserObjectInterface.h (hunk 1 above)
Comment (dschwen) [inline framework/include/userobject/UserObjectInterface.h]: Is any of that logic in the warehouse? Anyhow. These internals could be changed without changing the user facing interface.

Comment (dschwen) [inline framework/include/userobject/UserObjectInterface.h]: I see, you mean keep a separate list of the Discrete* objects in the warehouse? This should probably be done eventually. We could get rid of the dynamic casts on get that way and we could guarantee that execute etc. are never called. 

But given that we have three people with separate PRs waiting on this functionality I'd like to pin down the user facing interface first and rework the warehouse stuff later. I don't even know it it makes sense to work on that part right now, with the warehouse refactor in progress.

**Comment by friedmud** [comment]
Ok - I don't think there's anything else to change currently then.  Go ahead and to one final (hopefully!) rename...

**Comment by dschwen** [comment]
Ok, what did we agree on now? Back to MaterialUserObject?

**Comment by friedmud** [comment]
I think `DiscreteElementUserObject` will be fine.  It definitely sets it apart as something "different" that people won't associate with doing anything else ;-)

**Comment by dschwen** [comment]
Ok, changes are made!

**Comment by aeslaughter** [comment]
Don't worry about all the warehouse stuff, I am in the process of updating
UserObjects to the new design. I will take care of it.

**Comment by dschwen** [comment]
Awesome, thanks @aeslaughter !
